backend/perf.js

```diff
@@ -80,275 +79,285 @@ function init(repo_loc, callback) {
-                            // We're computing the last week, so let's also compute the total
-                            // difference.
-                            var difference = summary[kind][0].by_crate[c][p] - values[1];
-                            var percent = 100 * difference / values[1];
-                            summary[kind][0].by_crate[c][p] = percent;
-                        }
-                    } else if (i == 0) {
-                        // Save the first week we look at so we can compute the total difference,
-                        // see above.
-                        summary[kind][i].by_crate[c][p] = values[1]
-                    }
+    for (let crate_name in weeks[0]) {
+        if (!weeks[1][crate_name] || !weeks[2][crate_name]) {
+            continue;
+        }
+
+        by_crate[crate_name] = {};
+        let cur_summary = {};
+        for (let phase of Object.keys(weeks[0][crate_name])) {
+            if (!weeks[1][crate_name][phase] || !weeks[2][crate_name][phase]) {
+                continue;
+            }
+
+            // Find the median value.
+            let values = [
+                weeks[0][crate_name][phase].time,
+                weeks[1][crate_name][phase].time,
+                weeks[2][crate_name][phase].time
+            ];
+            values.sort((a, b) => a - b);
+            let median = values[1];
+
+            by_crate[crate_name][phase] = median;
+
+            // Previous summary summarized this kind, had this crate, and
+            // get it's value for the phase
+            let previous_median = summary_state[crate_name] &&
+                summary_state[crate_name][phase];
+
+            if (previous_median > 0) {
+                cur_summary[phase] =
+                    get_percent_change(previous_median, median);
+
+                if (i == WEEKS_IN_SUMMARY &&
+                    summary[kind][0].by_crate[crate_name] &&
+                    summary[kind][0].by_crate[crate_name][phase] &&
+                    summary[kind][0].by_crate[crate_name][phase] > 0) {
+                    // We're computing the last week, so let's also
+                    // compute the total difference.
+                    summary[kind][0].by_crate[crate_name][phase] =
+                        get_percent_change(summary[kind][0].by_crate[crate_name][phase], median);
                 }
+            } else if (i == 0) {
+                // Save the first week we look at so we can compute the
+                // total difference, see above.
+                cur_summary[phase] = median;
             }
         }
+
+        summary[kind][i].by_crate[crate_name] = cur_summary;
     }
 
-    prev_summary[kind] = cur_summary[kind];
+    return by_crate;
 }
 
 
 function make_data(date, header, times, rustc_crate) {
-    if (!header || !times || times.length == 0) {
-        return {};
-    }
```

Comment:
From what I can tell, none of these were ever possibly true. We check [and skip with `continue`] for `times.length == 0`; and both `header` and `times` are accessed previously, so they could never be false. If we feel this is a good sanity check though I can remove this change.